### ISSUE TITLE
adding batchnorm/dropout layers and setting training=True/False for training/eval actor respectively

### ISSUE THREAD

**Issue by turmeric-blend** [google-research/seed_rl#28]
Hi, I want to add batchnorm/dropout layers to my network but its unclear how to set the layer's ['training' argument](https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization#call_arguments_2) to True/False for the training/eval actors respectively.

**Comment by lespeholt** [comment]
Before adding batchnorm you need to consider what to do with actor/learner respectively as if you're not careful, actors will end up being more off policy than without batch norm. Other forms of normalization are usually easier to deal with like layer norm.

**Comment by turmeric-blend** [comment]
Hi @lespeholt , would you mind explaining more about how batchnorm would make actor more off policy?

**Comment by lespeholt** [comment]
If you apply batch norm naively, it won't necessarily work well. The inference batch and the training batch is different so different statistics will be applied and hence the actor and learner policy will differ more than usual. At inference you may have batch size 32 with 1 time step and on training you may have batch size 32 and 100 time steps. And the actors at inference and in the training batch are also with high likelihood different.

**Comment by turmeric-blend** [comment]
Hi @lespeholt I understand these concerns. I think I should clarify that I plan to implement batchnorm at `self._body` only.

My idea was that for training & inference on the same atari game, the training and inference batch won't be so different? Also my plan was to turn off batchnorm at eval as it is the standard practice, thus helping training convergence and won't influence eval too much, hence the requirement to set it True/False at training/eval. I think this should be the right way if wanting to implement batchnorm while reducing statistics effects?

Also for the 1 vs 100 time step at the LSTMCell (is this where you suggest the LayerNorm?), if inference was done on the most recent frame + 99 past frames, then this wouldn't be an issue?

**Comment by lespeholt** [comment]
This problem will also be there if you only do it for body. You can try it out and see what happens, but in case it doesn't work well, it's likely due to V-trace removing some of the signal as inference and training is further from each other. Will still say that any of the other normalization tricks are more straightforward and I know layer norm has helped. Group norm is also an option.

**Comment by turmeric-blend** [comment]
I realized I did not state it clearly at the start. I am using R2D2 instead of V-Trace. I am not familiar with V-Trace, but would any of your concerns still apply to R2D2?

Nevertheless, I am keen to try it all the norm methods out, however would appreciate some guidance on how to set it up for seed_rl. I am thinking along the lines of, for 4 actors (3 training & 1 eval), the 3 training actors would have norm(training=True) (including the normlayer) and the 1 eval actor have norm(training-False) (removing the normlayer).
